Signal/src/Models/TSMessageAdapaters/TSVideoAttachmentAdapter.m

```diff
@@ -144,79 +185,96 @@ - (UIView *)mediaView {
-        _durationLabel.font   = [UIFont systemFontOfSize:14];
-        [_durationLabel sizeToFit];
-        _durationLabel.frame = CGRectMake((size.width - _durationLabel.frame.size.width) - 10,
-                                          _durationLabel.frame.origin.y,
-                                          _durationLabel.frame.size.width,
-                                          AUDIO_BAR_HEIGHT);
-        _durationLabel.backgroundColor = [UIColor clearColor];
-        _durationLabel.textColor       = [UIColor whiteColor];
-
-        if (_incoming) {
-            audioBubble.backgroundColor =
-                [UIColor colorWithRed:229 / 255.0f green:228 / 255.0f blue:234 / 255.0f alpha:1.0f];
-            _waveform.normalColor = [UIColor whiteColor];
-            _waveform.progressColor =
-                [UIColor colorWithRed:107 / 255.0f green:185 / 255.0f blue:254 / 255.0f alpha:1.0f];
-            _durationLabel.textColor = [UIColor darkTextColor];
-        }
-
-        [audioBubble addSubview:_waveform];
-        [audioBubble addSubview:_audioPlayPauseButton];
-        [audioBubble addSubview:_durationLabel];
-
-        if (!_incoming) {
-            self.attachmentUploadView = [[AttachmentUploadView alloc] initWithAttachment:self.attachment
-                                                                               superview:audioBubble
-                                                                 attachmentStateCallback:nil];
+        if (self.cachedMediaView == nil) {
+            CGSize viewSize = [self mediaViewDisplaySize];
+            UIColor *textColor = [self audioTextColor];
+
+            _cachedMediaView = [[UIView alloc] initWithFrame:CGRectMake(0.f, 0.f, viewSize.width, viewSize.height)];
+
+            _cachedMediaView.backgroundColor
+                = self.incoming ? [UIColor jsq_messageBubbleLightGrayColor] : [UIColor ows_materialBlueColor];
+            [JSQMessagesMediaViewBubbleImageMasker applyBubbleImageMaskToMediaView:_cachedMediaView
+                                                                        isOutgoing:!self.incoming];
+
+            const CGFloat kBubbleTailWidth = 6.f;
+            CGRect contentFrame = CGRectMake(self.incoming ? kBubbleTailWidth : 0.f,
+                self.vMargin,
+                viewSize.width - kBubbleTailWidth - (self.incoming ? 10 : 15),
+                viewSize.height - self.vMargin * 2);
+
+            CGRect iconFrame = CGRectMake(round(contentFrame.origin.x + 10.f),
+                round(contentFrame.origin.y + (contentFrame.size.height - self.iconSize) * 0.5f),
+                self.iconSize,
+                self.iconSize);
+            _audioPlayPauseButton = [[UIButton alloc] initWithFrame:iconFrame];
+            _audioPlayPauseButton.enabled = NO;
+            [_cachedMediaView addSubview:_audioPlayPauseButton];
+
+            const CGFloat kLabelHSpacing = 3;
+            const CGFloat kLabelVSpacing = 2;
+            NSString *topText =
```

Comment:
Yes, I'd love to. So far I have resisted the urge to pull out this logic.

* We're likely to revisit all of this "soon" when we replace JSQ.
* It's not like we'll keep adding more attachment adapters.
* There are slight differences between the attachment adapter's appearance which would complicate this.
* The attachment approval dialog and the "full screen" attachment view are way larger; they won't re-use any of this view/layout logic.